### PR TITLE
Add fingerprint command for emulators. ApiLevel >= 23 required

### PR DESCRIPTION
**Pull request by vrunoa**
Add the possibility to execute ```adb emu finger touch <fingerprint_id>``` for emulator with **ApiLevel >=23**. The idea is to add a method on appium-android-driver, and then on appium, so users can test the fingerprint implementation with appium.

### CODE REVIEW

**Comment by imurchie** [comment]
Thanks for putting this together!

Can you run `gulp jshint` locally? Also there are failures in the build.

**Comment by vrunoa** [comment]
@imurchie here the ```gulp jshint``` output.
```
gulp jshint
[12:30:21] Using gulpfile ~/Urucas/Node/appium-adb/gulpfile.js
[12:30:21] Starting 'jshint'...
[12:30:22] Finished 'jshint' after 1.44 s
```
I'll check what's failing on travis and let you know, since the test are passing locally.

**Comment by vrunoa** [comment]
@imurchie fixed. I refactor ```.find``` to a simple ```for``` in ```isEmulatorConnected``` method. Didn't remember ```.find``` wasn't supported on Node **0.12.x**, I was working with **v5.5.0.**

I attach you the functional test,
```bash
gulp transpile && mocha -t 90000 -R spec build/test/functional/adb-emu-commands-e2e-specs.js
[12:04:55] Using gulpfile ~/Urucas/Node/appium-adb/gulpfile.js
[12:04:55] Starting 'transpile'...
[12:05:04] Finished 'transpile' after 8.44 s


  adb emu commands
info ADB Checking whether adb is present
info ADB Using adb from /Users/vrunoa/Library/Android/sdk/platform-tools/adb
dbug ADB Getting install status for com.example.fingerprint
dbug ADB Getting connected devices...
dbug ADB 1 device(s) connected
dbug ADB Running /Users/vrunoa/Library/Android/sdk/platform-tools/adb with args: ["-P",5037,"shell","getprop","ro.build.version.sdk"]
dbug ADB Device API level: 23
dbug ADB Getting connected devices...
dbug ADB 1 device(s) connected
dbug ADB Running /Users/vrunoa/Library/Android/sdk/platform-tools/adb with args: ["-P",5037,"shell","pm","list","packages","-3","com.example.fingerprint"]
dbug ADB App is  installed
dbug ADB Getting connected devices...
dbug ADB 1 device(s) connected
dbug ADB Running /Users/vrunoa/Library/Android/sdk/platform-tools/adb with args: ["-P",5037,"shell","am","force-stop","com.example.fingerprint"]
dbug ADB Uninstalling com.example.fingerprint
dbug ADB Getting connected devices...
dbug ADB 1 device(s) connected
dbug ADB Running /Users/vrunoa/Library/Android/sdk/platform-tools/adb with args: ["-P",5037,"shell","am","force-stop","com.example.fingerprint"]
dbug ADB Running /Users/vrunoa/Library/Android/sdk/platform-tools/adb with args: ["-P",5037,"uninstall","com.example.fingerprint"]
info ADB App was uninstalled
dbug ADB Running /Users/vrunoa/Library/Android/sdk/platform-tools/adb with args: ["-P",5037,"install","-r","/Users/vrunoa/Urucas/Node/appium-adb/test/fixtures/Fingerprint.apk"]
dbug ADB Device API level: 23
dbug ADB Getting connected devices...
dbug ADB 1 device(s) connected
dbug ADB Running /Users/vrunoa/Library/Android/sdk/platform-tools/adb with args: ["-P",5037,"shell","am","start","-n","com.example.fingerprint/.MainActivity","-S"]
dbug ADB Getting focused package and activity
dbug ADB Getting connected devices...
dbug ADB 1 device(s) connected
dbug ADB Running /Users/vrunoa/Library/Android/sdk/platform-tools/adb with args: ["-P",5037,"shell","dumpsys","window","windows"]
dbug ADB Getting connected emulators
dbug ADB Getting connected devices...
dbug ADB 1 device(s) connected
dbug ADB 1 emulator(s) connected
dbug ADB Setting device id to emulator-5554
dbug ADB Device API level: 23
dbug ADB Running /Users/vrunoa/Library/Android/sdk/platform-tools/adb with args: ["-P",5037,"-s","emulator-5554","emu","finger","touch",1111]
dbug ADB Getting focused package and activity
dbug ADB Getting connected devices...
dbug ADB 1 device(s) connected
dbug ADB Running /Users/vrunoa/Library/Android/sdk/platform-tools/adb with args: ["-P",5037,"-s","emulator-5554","shell","dumpsys","window","windows"]
    ✓ fingerprint should close app on emitted finger touch event (15921ms)


  1 passing (16s)
```
for this; the emulator api level must be >23 and also it must have a fingerprint defined in security settings.  I'll search for an alternative, maybe just wait for the fingerprint event and don't validate against the one defined in security.

**Comment by imurchie** [comment]
Still some problem with launching the activity.

Also: where does the fingerprint app come from? We don't want to have a random app in our codebase with no understanding of it.

**Comment by vrunoa** [comment]
@imurchie here's the source code for the fingerprint app; [https://github.com/Urucas/android-fingerprint-example[(https://github.com/Urucas/android-fingerprint-example). In case you feel it's important I can transfer the project ownership to you guys and then edit or create PR to that project. 

I can't find the **e2e-Tests** results so I can fix them. Can you give me a hand on this ?

**Comment by imurchie** [comment]
Ok! We should add something to the README about the source. I'm fine with it not being in the Appium repo for now. 

Console logs from the e2e tests: https://team-appium.ci.cloudbees.com/job/adb-e2e-android/117/console

In case it is still not visible,

```
adb emu commands
info ADB Checking whether adb is present
info ADB Using adb from /home/appium/Tools/android-sdk-linux/platform-tools/adb
dbug ADB Getting install status for com.example.fingerprint
dbug ADB Getting connected devices...
dbug ADB 1 device(s) connected
dbug ADB Running /home/appium/Tools/android-sdk-linux/platform-tools/adb with args: ["-P",5037,"shell","getprop","ro.build.version.sdk"]
dbug ADB Device API level: 18
dbug ADB Getting connected devices...
dbug ADB 1 device(s) connected
dbug ADB Running /home/appium/Tools/android-sdk-linux/platform-tools/adb with args: ["-P",5037,"shell","pm","list","packages","-3","com.example.fingerprint"]
dbug ADB App is  not installed
dbug ADB Running /home/appium/Tools/android-sdk-linux/platform-tools/adb with args: ["-P",5037,"install","-r","/home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/test/fixtures/Fingerprint.apk"]
dbug ADB Device API level: 18
dbug ADB Getting connected devices...
dbug ADB 1 device(s) connected
dbug ADB Running /home/appium/Tools/android-sdk-linux/platform-tools/adb with args: ["-P",5037,"shell","am","start","-n","com.example.fingerprint/.MainActivity","-S"]
ERR! ADB Error: Activity used to start app doesn't exist or cannot be launched! Make sure it exists and is a launchable activity
ERR! ADB     at Object.wrappedLogger.errorAndThrow (lib/logger.js:60:13)
ERR! ADB     at ADB.callee$0$0$ (lib/tools/apk-utils.js:68:13)
ERR! ADB     at tryCatch (/home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/babel-runtime/regenerator/runtime.js:67:40)
ERR! ADB     at GeneratorFunctionPrototype.invoke [as _invoke] (/home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/babel-runtime/regenerator/runtime.js:315:22)
ERR! ADB     at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/babel-runtime/regenerator/runtime.js:100:21)
ERR! ADB     at GeneratorFunctionPrototype.invoke (/home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/babel-runtime/regenerator/runtime.js:136:37)
ERR! ADB     at bound (domain.js:287:14)
ERR! ADB     at GeneratorFunctionPrototype.runBound (domain.js:300:12)
ERR! ADB     at run (/home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/core-js/library/modules/es6.promise.js:104:47)
ERR! ADB     at /home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/core-js/library/modules/es6.promise.js:115:28
ERR! ADB     at flush (/home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/core-js/library/modules/$.microtask.js:19:5)
ERR! ADB     at nextTickCallbackWith0Args (node.js:453:9)
ERR! ADB     at process._tickDomainCallback (node.js:423:13)
ERR! ADB  [Error: Activity used to start app doesn't exist or cannot be launched! Make sure it exists and is a launchable activity]
ERR! ADB Error: Error occured while starting App. Original error: Activity used to start app doesn't exist or cannot be launched! Make sure it exists and is a launchable activity
ERR! ADB     at Object.wrappedLogger.errorAndThrow (lib/logger.js:60:13)
ERR! ADB     at ADB.callee$0$0$ (lib/tools/apk-utils.js:80:9)
ERR! ADB     at tryCatch (/home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/babel-runtime/regenerator/runtime.js:67:40)
ERR! ADB     at GeneratorFunctionPrototype.invoke [as _invoke] (/home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/babel-runtime/regenerator/runtime.js:315:22)
ERR! ADB     at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/babel-runtime/regenerator/runtime.js:100:21)
ERR! ADB     at GeneratorFunctionPrototype.invoke (/home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/babel-runtime/regenerator/runtime.js:136:37)
ERR! ADB     at bound (domain.js:287:14)
ERR! ADB     at GeneratorFunctionPrototype.runBound (domain.js:300:12)
ERR! ADB     at run (/home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/core-js/library/modules/es6.promise.js:104:47)
ERR! ADB     at /home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/core-js/library/modules/es6.promise.js:115:28
ERR! ADB     at flush (/home/appium/Workspace/jenkins-remote-executor/workspace/adb-e2e-android/node_modules/core-js/library/modules/$.microtask.js:19:5)
ERR! ADB     at nextTickCallbackWith0Args (node.js:453:9)
ERR! ADB     at process._tickDomainCallback (node.js:423:13)
ERR! ADB  [Error: Error occured while starting App. Original error: Activity used to start app doesn't exist or cannot be launched! Make sure it exists and is a launchable activity]

    1) fingerprint should open the secret activity on emitted valid finger touch event
```

**Comment by vrunoa** [comment]
Cool @imurchie it's visible. I'm seeing the installation failed since the Api Level in the emulator is 18, and the app minimum required level is 23.
```
dbug ADB App is  not installed
```
I believe that's why ```appium-adb``` can't start the app. What do you think would be the best solution for this case?

**Comment by moizjv** [comment]
@vrunoa may be you can skip test for CI and add a comment or even better throw not supported exception if API level is < 23, what do you think @imurchie

File: lib/tools/adb-emu-commands.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        log.errorAndThrow(`No emulators available.`);
+        return;
+      }
+      udid = emulators[0].udid;
+    }
+    await this.setDeviceId(udid);
+    let level = await this.getApiLevel();
+    if((~~level) < 23) {
+      log.errorAndThrow(`Emulator Api Level must be >= 23. Current Api level ${level}`);
+      return;
```
Comment (moizjv) [inline lib/tools/adb-emu-commands.js]: i dont think we need return here

Comment (vrunoa) [inline lib/tools/adb-emu-commands.js]: Perfect. I'll remove them.

File: lib/tools/adb-emu-commands.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    if(emulators[i].udid === udid) return true;
+  }
+  return false;
+};
+
+emuMethods.fingerprint = async function(fingerprint_id, udid = undefined) {
+  try {
+    if(!fingerprint_id) {
+      log.errorAndThrow(`Fingerprint id param must be defined`);
+      return;
```
Comment (moizjv) [inline lib/tools/adb-emu-commands.js]: here too can remove return

File: lib/tools/adb-emu-commands.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    }
+    await this.setDeviceId(udid);
+    let level = await this.getApiLevel();
+    if((~~level) < 23) {
+      log.errorAndThrow(`Emulator Api Level must be >= 23. Current Api level ${level}`);
+      return;
+    }
+    await this.adbExec(["emu", "finger", "touch", fingerprint_id]);
+    return true;
+  }catch(e){
```
Comment (moizjv) [inline lib/tools/adb-emu-commands.js]: `} catch(e) {`

File: lib/tools/adb-emu-commands.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    log.errorAndThrow(`Error getting emulators. Original error: ${e.message}`);
+  }
+};
+
+export default emuMethods;
```
Comment (moizjv) [inline lib/tools/adb-emu-commands.js]: looks like editor has added some thing in the end

File: lib/tools/system-calls.js [hunk 5]
```diff
@@ -271,6 +271,10 @@ systemCallMethods.setEmulatorPort = function (emPort) {
 systemCallMethods.setDeviceId = function (deviceId) {
   log.debug(`Setting device id to ${deviceId}`);
   this.curDeviceId = deviceId;
+  let argsHasDevice = this.executable.defaultArgs.indexOf("-s"); 
+  if(argsHasDevice !== -1) {
```
Comment (moizjv) [inline lib/tools/system-calls.js]: `if (argsHasDevice !== -1) { `

File: test/functional/adb-emu-commands-e2e-specs.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  activity = '.MainActivity',
+  secretActivity = '.SecretActivity';
+  
+describe('adb emu commands', () => {
+  let adb;
+  before(async () => {
+    adb = await ADB.createADB();
+  });
+  it('fingerprint should open the secret activity on emitted valid finger touch event', async () => {
+    if(await adb.isAppInstalled(pkg)) {
```
Comment (moizjv) [inline test/functional/adb-emu-commands-e2e-specs.js]: `if (`

File: test/unit/adb-emu-commands-specs.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+describe('adb emulator commands', () => {
+  let adb = new ADB();
+  describe("emu", () => {
+    describe("isEmulatorConnected", withMocks({adb}, (mocks) => {
+      it("should check emulator is connected", async () => {
+        mocks.adb.expects("getConnectedEmulators")
+          .atLeast(1).withExactArgs()
+          .returns(emulators);
+        (await adb.isEmulatorConnected("emulator-5554")).should.equal(true);
```
Comment (moizjv) [inline test/unit/adb-emu-commands-specs.js]: i think you can use `await adb.isEmulatorConnected("emulator-5554").should.eventually.equal(true);`

File: test/unit/adb-emu-commands-specs.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+          .atLeast(1).withExactArgs()
+          .returns("23");
+        mocks.adb.expects("setDeviceId")
+          .atLeast(1).withExactArgs("emulator-5554")
+          .returns();
+        mocks.adb.expects("adbExec")
+          .atLeast(1)
+          .withExactArgs(["emu", "finger", "touch", fingerprint])
+          .returns("");
+        chai.expect(await adb.fingerprint(fingerprint)).to.be.true;
```
Comment (moizjv) [inline test/unit/adb-emu-commands-specs.js]: why not use `should.eventually` to be consistent

Comment (vrunoa) [inline test/unit/adb-emu-commands-specs.js]: done it. 
https://github.com/vrunoa/appium-adb/blob/master/test/unit/adb-emu-commands-specs.js#L50

File: .travis.yml [hunk 9]
```diff
@@ -1,5 +1,5 @@
 language: node_js
-node_js:
+node_js: 
   - "0.12"
-after_success:
+after_sucess:
```
Comment (imurchie) [inline .travis.yml]: Why?

**Comment by imurchie** [comment]
Merging. I'll do some little cleanup afterwards.

**Comment by vrunoa** [comment]
Cool! I'll add the current method in the ```appium-android-driver``` next week.

File: .travis.yml (hunk 9 above)
Comment (vrunoa) [inline .travis.yml]: Honestly, I don't know what happened there.

**Pull request merged**